circus-train-api/src/main/java/com/hotels/bdp/circustrain/api/metastore/CloseableMetaStoreClientFactory.java

```diff
@@ -36,20 +42,48 @@ public static CloseableMetaStoreClient newInstance(IMetaStoreClient delegate) {
   static class CloseableMetaStoreClientInvocationHandler implements InvocationHandler {
 
     private final IMetaStoreClient delegate;
+    private HiveMetaStoreClientCompatibility12x compatibility;
 
     CloseableMetaStoreClientInvocationHandler(IMetaStoreClient delegate) {
       this.delegate = delegate;
+      try {
+        compatibility = new HiveMetaStoreClientCompatibility12x(delegate);
+      } catch (Throwable t) {
+        log.warn("Unable to initialize compatibility", t);
+      }
     }
 
     @Override
     public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
       try {
         return method.invoke(delegate, args);
       } catch (InvocationTargetException e) {
+        try {
+          if (compatibility != null && e.getCause().getClass().isAssignableFrom(TApplicationException.class)) {
+            return invokeCompatibility(method, args);
+          }
+        } catch (Throwable t) {
+          log.warn("Unable to run compatibility for metastore client method {}. Will rethrow original exception: ",
+              method.getName(), t);
+        }
         throw e.getCause();
       }
     }
 
+    private Object invokeCompatibility(Method method, Object[] args) throws Throwable {
+      Class<?>[] argTypes = getTypes(args);
+      Method compatibilityMethod = compatibility.getClass().getMethod(method.getName(), argTypes);
+      return compatibilityMethod.invoke(compatibility, args);
+    }
+
+    private static Class<?>[] getTypes(Object[] args) {
```

Comment:
Where does this code come from? Some third party library right? If so we should reference it here, update the license header above and also add this class to the list of files which the license plugin doesn't automatically update.

Reply:
this code is mine

Reply:
OK, loving the performance enhancing ++i as opposed to i++ ;p